pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -137,6 +151,16 @@ func (rc *reconciler) reconcile() {
 	for _, attachedVolume := range rc.actualStateOfWorld.GetAttachedVolumes() {
 		if !rc.desiredStateOfWorld.VolumeExists(
 			attachedVolume.VolumeName, attachedVolume.NodeName) {
+
+			// Don't even try to start an operation if there is already one running
+			// This check must be done before we do any other checks, as otherwise the other checks
+			// may pass while at the same time the volume leaves the pending state, resulting in
+			// double detach attempts
+			if rc.attacherDetacher.IsOperationPending(attachedVolume.VolumeName, "") {
```

Comment:
But this checking still cannot guarantee detach only happen once because of asynchronous operation. For example, when it checks IsOperationPending in line 159, the first detach operation just finishes, so it could pass the check and continue to issue detach operation a second time. I think the test should not assume detach operation happens exactly once.